rust/kernel/sync/arc.rs

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! A reference-counted pointer.
+//!
+//! This module implements a way for users to create reference-counted objects and pointers to them
+//! that automatically increment and decrement the count and drops the underlying object when it
+//! reaches zero.
+//!
+//! It is different from the standard library's `Arc` in two ways: it does not support weak
+//! references, which allows it to be smaller -- a single pointer-sized integer; it allows users to
+//! safely increment the reference count from a single reference to the underlying object.
+
+use crate::KernelResult;
+use alloc::boxed::Box;
+use core::{
+    mem::ManuallyDrop,
+    ops::Deref,
+    ptr::NonNull,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+/// A reference-counted pointer to an instance of `T`.
+///
+/// The reference count is incremented when new instances of [`Ref`] are created, and decremented
+/// when they are dropped. When the count reaches zero, the underlying `T` is also dropped.
+///
+/// # Invariants
+///
+/// The value stored in [`T::get_count`] corresponds to the number of instances of [`Ref`] that
+/// point to that instance of `T`.
+pub struct Ref<T: RefCounted + ?Sized> {
+    ptr: NonNull<T>,
+}
+
+impl<T: RefCounted> Ref<T> {
+    /// Constructs a new reference counted instance of `T`.
+    pub fn try_new(contents: T) -> KernelResult<Self> {
+        let boxed = Box::try_new(contents)?;
+        boxed.get_count().count.store(1, Ordering::Release);
+        let ptr = NonNull::from(boxed.deref());
+        Box::into_raw(boxed);
+        Ok(Ref { ptr })
+    }
+}
+
+impl<T: RefCounted + ?Sized> Ref<T> {
+    /// Creates a new reference-counted pointer to the given instance of `T`. It works by
+    /// incrementing the current reference count as part of constructing the new pointer.
```

Comment:
The first sentence is special, like in Git commits, i.e. details go afterwards:

```suggestion
    /// Creates a new reference-counted pointer to the given instance of `T`.
    ///
    /// It works by incrementing the current reference count as part of constructing
    /// the new pointer.
```

Reply:
I see this being done in the standard library for modules and types, but I see lots of functions where this isn't followed (example [here](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.leak)).

Are we creating a new convention?

Reply:
Hmm... Perhaps it is only supposed to be done for modules and types but not methods (since e.g. modules show the first line in the "containing" page of the generated docs).

Nevertheless, I think it is a good convention to have: simpler to remember if we always do it, forces the writer to think about what is the most relevant information to put first, can be used to "fold" the docs in the generated pages (i.e. imagine a pre-folded docs on load except for the first page), etc.